ArchivesSpace integration

mistydemeo opened the pull request:
This makes quite a number of changes to the Archivists' Toolkit workflow and introduces ArchivesSpace as an option.
- Breaks out the existing ATK code from multiple locations into a single client class
- Introduces a new ArchivesSpace class with an identical interface
- Adds an ArchivesSpace upload script
- Separates the generic pair matching code from ATK-specific code; the new `pair_matcher` module is used by the `views_atk` module and the new `views_as` module
- Allows re-entering the pair matching UI
- Display existing matches when loading the pair matching UI
- Streamline the pair matcher by POSTing and DELETEing matches as soon as the "pair" button is hit, and remove the "save" button

Commit:
Archivist's Toolkit: separate code into client class

refs #7598

Commit:
Add ArchivesSpaceClient class

refs #7589

Commit:
Add ArchivesSpace upload script

refs #7595

Commit:
Add ArchivesSpace configuration form

refs #7598

Commit:
Add ArchivesSpace configuration to the admin sidebar

refs #7598

Commit:
main.models: add ArchivesSpaceDIPObjectResourcePairing for ArchivesSpace

refs #7598

Commit:
Add ArchivesSpace DIP upload choice

refs #7595

Commit:
Separate out matcher backend; add AS views

refs #7598

Commit:
ArchivesSpace: automatically visit matching UI

Commit:
Ingest: add button to revisit matching GUIs

refs #8101

Commit:
Add link to parent ID when viewing components

refs #8109

Commit:
AS matcher: display existing matches

refs #8101

Commit:
AS/ATK matcher: remove save button

refs #8101

Commit:
AS/ATK matcher: update server state with every pair

refs #8101

Commit:
AS/ATK matcher: link to resource parent if appropriate

refs #8109

Commit:
AS/ATK: fetch all resources, not just collections

Commit:
Pair matcher: set page size to 30

Commit:
AS/ATK matcher: add "select all" file button

Commit:
Add "Finish matching" button to redirect to Ingest

Commit:
Pair matcher: correctly page results

Previous to this, the Archivists' Toolkit code would fetch all results
from the database. This is relatively wasteful, but is *much* slower in
ArchivesSpace due to the large numbers of REST calls.

This instead fetches as few records as possible in order to improve
speed. In order to allow the Django Paginator to still be used, a new
list-like type has been introduced which provides a list-like interface
to paged data sources like REST clients.

Commit:
Pair matcher: implement sorting of resources

Commit:
Pair matcher: count collections using given query

Commit:
Pair matcher: fix matching multiple objects

Commit:
Pair matcher: add reset button

Commit:
AS: display component identifier, not ID

Hwesta commented on src/MCPClient/lib/clientScripts/upload-archivesspace.py:
```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/python
```
Source compatible? :D

mistydemeo replied:
Probably not! This is the same shebang as upload-atk though... which probably wasn't intentional there

Hwesta replied:
Update this to /usr/bin/env python2 to match #255 ?

Hwesta commented on src/MCPClient/lib/clientScripts/upload-archivesspace.py:
```diff
@@ -0,0 +1,169 @@
+    # get a directory listing
+    # for each item, set fileName and go
+    try:
+        mydir = dip_location + "objects/"
```
os.path.join?  Or is this a URL?

Hwesta commented on src/MCPClient/lib/clientScripts/upload-archivesspace.py:
```diff
@@ -0,0 +1,169 @@
+        else:
+            logger.error("no files in " + mydir)
+            raise ValueError("cannot find dip")
+            exit(2)
```
Will exit ever get called? Won't the raise trigger first and end the function?  Same below.

Hwesta commented on src/MCPClient/lib/clientScripts/upload-archivesspace.py:
```diff
@@ -0,0 +1,169 @@
+
+
+def upload_to_archivesspace(files, client, xlink_show, xlink_actuate, object_type, use_statement, uri_prefix, dip_uuid, access_conditions, use_conditions, restrictions, dip_location):
+
```
Docstring?  Also elsewhere, but holy cow this has a lot of params.

mistydemeo replied:
Yeah, a _lot_ of the structure of this is copied from upload-atk. :cold_sweat:

Hwesta replied:
Ahh, I didn't realize that (I didn't see a matching removal diff). Docstring extra useful then, if extra painful to write. :P

mistydemeo replied:
The other script still exists; I used it as a template, since this does mostly the same things with some customizations. Eventually might want to refactor into one script with two backends but that wasn't feasible at this point in time.

Hwesta commented on src/MCPClient/lib/clientScripts/upload-archivesspace.py:
```diff
@@ -0,0 +1,169 @@
+import os
+from urlparse import urljoin
+
+from main.models import ArchivesSpaceDIPObjectResourcePairing
```
Depending on the order that #124 comes in this might have to be updated.

Hwesta commented on src/MCPClient/lib/clientScripts/upload-archivesspace.py:
```diff
@@ -0,0 +1,169 @@
+    mets = None
+    if restrictions == 'premis' or len(access_conditions) == 0 or len(use_conditions) == 0:
+        logger.debug("Looking for mets: {}".format(dip_uuid))
+        mets_source = dip_location + 'METS.' + dip_uuid + '.xml'
```
os.path.join?  Are there any guarantees on dip_location?

Hwesta commented on src/archivematicaCommon/lib/archivesspace/client.py:
```diff
@@ -0,0 +1,424 @@
+    pass
+
+
+class CommunicationError(Exception):
```
Should these inherit from `ArchivesSpaceError`? So it's easier to catch anything generated by this module?

mistydemeo replied:
Good thought.

Hwesta commented on src/archivematicaCommon/lib/archivesspace/client.py:
```diff
@@ -0,0 +1,424 @@
+
+        try:
+            output = response.json()
+        except:
```
Catch the specific error(s)? ValueError?

mistydemeo replied:
It looks like the raised exception is a `simplejson.scanner.JSONDecodeError`, but I'm not 100% sure if requests always uses that as a JSON decoder?

mistydemeo replied:
Looks like the surface area where this could come up is limited, but it's possible for the JSON decoder to not be simplejson (and hence for the exception to be a different class). Think I'll leave this alone.

mistydemeo replied:
https://github.com/kennethreitz/requests/blob/8b5e457b756b2ab4c02473f7a42c2e0201ecc7e9/requests/compat.py#L24-L29

Hwesta replied:
Then how about except Exception:  so we don't catch SysExit etc

mistydemeo replied:
Updated!

Hwesta commented on src/archivematicaCommon/lib/archivesspace/client.py:
```diff
@@ -0,0 +1,424 @@
+    pass
+
+
+class ArchivesSpaceClient(object):
```
Tests for this?

Hwesta commented on src/archivematicaCommon/lib/archivesspace/client.py:
```diff
@@ -0,0 +1,424 @@
+            # session has expired; acquire a new session, then retry once
+            if retry and response.status_code == 412 and "SESSION_GONE" in response.text:
+                self._login()
+                return self._request(method, url, params, expected_response, data=data, retry=False)
```
Could this infinitely recurse if the session isn't created each time? (Is that out of scope?)

mistydemeo replied:
It recurses with the `retry` kwarg set to `False`, and L78 checks that first - it will never retry more than once.

Hwesta replied:
Ah, I didn't catch that.

Hwesta commented on src/archivematicaCommon/lib/archivesspace/client.py:
```diff
@@ -0,0 +1,424 @@
+                return self._request(method, url, params, expected_response, data=data, retry=False)
+            raise CommunicationError(response.status_code, response)
+
+        output = response.json()
```
Is this likely to have a JSON parsing error?

Hwesta commented on src/archivematicaCommon/lib/archivesspace/client.py:
```diff
@@ -0,0 +1,424 @@
+                    kwargs = {'reverse': True} if sort_by == 'desc' else {}
+                    result['children'] = sorted(result['children'], key=lambda c: c['title'], **kwargs)
+            else:
+                result['children'] = False
```
None is a better indicator of no-data.

mistydemeo replied:
This is identical to the behaviour used by the previous Archivist's Toolkit code - I decided not to make any changes to the data format.

Hwesta commented on src/archivematicaCommon/lib/archivesspace/client.py:
```diff
@@ -0,0 +1,424 @@
+        :return tuple: A tuple containing:
+            * The type of the parent record; valid values are ArchivesSpaceClient.RESOURCE and ArchivesSpaceClient.RESOURCE_COMPONENT.
+            * The URL of the parent record.
+            If the provided URL fragment references a resource, this method will simpl return the same URL.
```
Typo: s/simpl/simply/

Commit:
AS matcher: add “review matches” screen

Commit:
AS: add "review matches" to matching screen

Commit:
AS/ATK: remove “finish matching” link except on “review matches”

Commit:
AS/ATK: restructure DIP URL

Commit:
AS: publish file versions if there are no restrictions

Commit:
AS: don't display resource ID if not present

Hwesta commented on src/archivematicaCommon/lib/lazy_paged_sequence.py:
```diff
@@ -0,0 +1,48 @@
+class LazyPagedSequence(object):
```
Do we want to vendor this or use the dependency lazy-paged-sequence?

mistydemeo replied:
Let's use the dependency; this is only vendored because I hadn't published it yet the last time this was touched.

Commit:
ArchivesSpaceClient: add tests

mistydemeo commented:
Added tests to `ArchivesSpaceClient`. They aren't 100% comprehensive but they cover most functionality.

Commit:
Separate AS and ATK matcher JavaScript

Reserve the new functionality for the ArchivesSpace path; keep copies
of the scripts under the ATK matcher name, which behave the same as
before.

mistydemeo commented:
Pushed a fix to the ATK path.

I think this is ready to be merged?

Hwesta commented on src/archivematicaCommon/tests/test_archivesspace_client.py:
```diff
@@ -0,0 +1,124 @@
+# -*- coding: UTF-8 -*-
```
:tada:  tests!

Hwesta commented:
:+1:

mistydemeo commented:
Merged! :fireworks:
